File: code/com/leetcode/solutions/NO_0004_寻找两个有序数组的中位数/Solution.js
/**
 * 思路：根据中位数的特性可知中位数左右两侧数字个数相等
 *      故只需猜测数组A中中位数位置，即可得出数组B中相应中位数位置
 *      使用二分法不断缩小数组A的中位数范围，最后求得中位数，时间复杂度为 O(log(m + n))
 * @param {number[]} nums1
 * @param {number[]} nums2
 * @return {number}
 */
var findMedianSortedArrays = function (A, B) {
    let m = A.length, n = B.length, isOdd = (m + n) % 2, res;
    if (m > n) {
        [m, n] = [n, m];
        [A, B] = [B, A];
    }
    const halfLen = ~~((m + n + 1) / 2);// j = halfLen - i;
    let imin = 0, imax = m;
    let i = (imin + imax) >> 1, j = halfLen - i;
    while (imin <= imax) {
        if (i < imax && B[j - 1] > A[i]) { //  A[i] 太小
            imin = i + 1;
        } else if (i > imin && B[j] < A[i - 1]) {//  A[i] 太大
            imax = i - 1;
        } else {
            res = isOdd ? Math.maxU(A[i - 1], B[j - 1]) : (Math.maxU(A[i - 1], B[j - 1]) + Math.minU(A[i], B[j])) / 2;
            break;
        }
        i = (imin + imax) >> 1, j = halfLen - i;
    }
    return res;
};

Math.maxU = (a, b) => {
    if (undefined === a) return b;
    if (undefined === b) return a;
    return Math.max(a, b);
}

Math.minU = (a, b) => {
    if (undefined === a) return b;
    if (undefined === b) return a;
    return Math.min(a, b);
}

File: code/com/leetcode/solutions/NO_0004_寻找两个有序数组的中位数/Solution.cs
using System;

public class Solution
{

    /*
   * nums1,  nums2:指定的有序数组
   * 思路：将两个数组合并为一个数组然后排序，求出中位数即可。
   * 其他：遍历连个数组合并为一个有序新数组，遍历的同时即排好序
  */
    public double FindMedianSortedArrays(int[] nums1, int[] nums2)
    {

        double ret = 0.0;
        int[] vs = new int[nums1.Length + nums2.Length];
        nums1.CopyTo(vs, 0);
        nums2.CopyTo(vs, nums1.Length);
        Array.Sort(vs);
        int y = vs.Length % 2;
        int index = vs.Length / 2;
        if (y == 0)
        {
            var d = (decimal)(vs[index] + vs[index - 1]) / 2;
            ret = (double)d;
        }
        else
        {
            ret = vs[index];
        }
        return ret;
    }


    public double FindMedianSortedArrays2(int[] nums1, int[] nums2)
    {
        double ret = 0.0;
        int[] vs = new int[nums1.Length + nums2.Length];
        if (nums1.Length > 0 && nums2.Length > 0)
        {
            int i = 0, j = 0, k = 0;
            bool n1 = true, n2 = true;
            while (k < vs.Length)
            {
                if (n1 && n2)
                {
                    if (nums1[i] < nums2[j])
                    {
                        vs[k] = nums1[i];
                        i++;
                        if (i == nums1.Length) { n1 = false; }
                    }
                    else
                    {
                        vs[k] = nums2[j];
                        j++;
                        if (j == nums2.Length) { n2 = false; };
                    }
                }
                else
                {
                    if (n1)
                    {
                        vs[k] = nums1[i];
                        i++;
                        if (i == nums1.Length) { n1 = false; }
                    }
                    else
                    {
                        vs[k] = nums2[j];
                        j++;
                        if (j == nums2.Length) { n2 = false; };
                    }
                }

                k++;
            }
        }
        else
        {
            if (nums1.Length == 0)
            {
                nums2.CopyTo(vs, 0);
            }
            else
            {
                nums1.CopyTo(vs, 0);
            }
        }
        int y = vs.Length % 2;
        int index = vs.Length / 2;
        if (y == 0)
        {
            var d = (decimal)(vs[index] + vs[index - 1]) / 2;
            ret = (double)d;
        }
        else
        {
            ret = vs[index];
        }
        return ret;



    }










}
